lib/streamlit/DeltaGenerator.py

```diff
@@ -1365,11 +1373,27 @@ def video(self, element, data, format="video/mp4", start_time=0):
         # TODO: Provide API to convert raw NumPy arrays to video file (with
         # proper headers, etc)?
         import streamlit.elements.generic_binary_proto as generic_binary_proto
+        from validators import url
 
-        generic_binary_proto.marshall(element.video, data)
         element.video.format = format
         element.video.start_time = start_time
 
+        if isinstance(data, string_types) and url(data):
+            # If this is a straight YouTube link, reformat it into an embed link
+            # before it gets shipped to the frontend.
+            # 
+            # Regex covers any youtube URL (incl. shortlinks) that points to a
+            # valid video but needs to be converted to an "embed" link.
+            re_yt_watchlink = re.compile(
+                "http(?:s?):\/\/(?:www\.)?youtu(?:be\.com\/watch\?v=|\.be\/)(?P<code>[\w\-\_]*)(&(amp;)?[\w\?=]*)?")
+            match = re_yt_watchlink.match(data)
+            if match:
+                element.video.url = "https://www.youtube.com/embed/{code}".format(**match.groupdict())
```

Comment:
Move the youtube pattern-matcher into a utility function in another file. If you look around and don't find a better place for it, add to `util.py`.

(At some point we should start breaking `util.py` into other utility files, grouped by subject area. Right now that file is a kitchen sink...)

Reply:
Yeah I was reticent to add yet another utility function there.  :)